scipy/linalg/cython_blas.pyx.in

```diff
@@ -0,0 +1,504 @@
+    'ctrmm',
+    'ctrmv',
+    'daxpy',
+    'dcopy',
+    'dgemm',
+    'dgemv',
+    'dger',
+    'drot',
+    'drotg',
+    'drotm',
+    'drotmg',
+    'dscal',
+    'dswap',
+    'dsymm',
+    'dsymv',
+    'dsyr',
+    'dsyr2',
+    'dsyr2k',
+    'dsyrk',
+    'dtrmm',
+    'dtrmv',
+    'saxpy',
+    'scopy',
+    'sgemm',
+    'sgemv',
+    'sger',
+    'srot',
+    'srotg',
+    'srotm',
+    'srotmg',
+    'sscal',
+    'sswap',
+    'ssymm',
+    'ssymv',
+    'ssyr',
+    'ssyr2',
+    'ssyr2k',
+    'ssyrk',
+    'strmm',
+    'strmv',
+    'zaxpy',
+    'zcopy',
+    'zdrot',
+    'zdscal',
+    'zgemm',
+    'zgemv',
+    'zgerc',
+    'zgeru',
+    'zhemm',
+    'zhemv',
+    'zher',
+    'zher2',
+    'zher2k',
+    'zherk',
+    'zrotg',
+    'zscal',
+    'zswap',
+    'zsymm',
+    'zsyr',
+    'zsyr2k',
+    'zsyrk',
+    'ztrmm',
+    'ztrmv',
+]
+
+}}
+
+{{for sig in WRAPPED_FUNCTIONS}}
+{{wrap_fortran_function(sig)}}
+{{endfor}}
+
+{{for name in WRAPPED_SUBROUTINES}}
+{{wrap_fortran_subroutine(name)}}
+{{endfor}}
+
+
+#
+# Friendly wrappers
+#
+
+cdef inline bint is_contiguous(double[:,:] a, int axis):
+    return (a.strides[axis] == sizeof(a[0,0]) or a.shape[axis] == 1)
+
+cpdef float complex cdotc(float complex[:] cx, float complex[:] cy):
+    cdef:
+        int n = cx.shape[0]
+        int incx = cx.strides[0] // sizeof(cx[0])
+        int incy = cy.strides[0] // sizeof(cy[0])
+    return cdotc_f(&n, &cx[0], &incx, &cy[0], &incy)
+
+cpdef float complex cdotu(float complex[:] cx, float complex[:] cy):
+    cdef:
+        int n = cx.shape[0]
+        int incx = cx.strides[0] // sizeof(cx[0])
+        int incy = cy.strides[0] // sizeof(cy[0])
+    return cdotu_f(&n, &cx[0], &incx, &cy[0], &incy)
+
+cpdef double dasum(double[:] dx):
+    cdef:
+        int n = dx.shape[0]
+        int incx = dx.strides[0] // sizeof(dx[0])
+    return dasum_f(&n, &dx[0], &incx)
+
+cpdef double ddot(double[:] dx, double[:] dy):
+    cdef:
+        int n = dx.shape[0]
+        int incx = dx.strides[0] // sizeof(dx[0])
+        int incy = dy.strides[0] // sizeof(dy[0])
+    return ddot_f(&n, &dx[0], &incx, &dy[0], &incy)
+
+cpdef int dgemm(double alpha, double[:,:] a, double[:,:] b, double beta,
+                double[:,:] c) except -1:
+    cdef:
+        char *transa
+        char *transb
+        int m, n, k, lda, ldb, ldc
+        double *a0=&a[0,0]
+        double *b0=&b[0,0]
+        double *c0=&c[0,0]
+    # In the case that c is C contiguous, swap a and b and
+    # swap whether or not each of them is transposed.
+    # This can be done because a.dot(b) = b.T.dot(a.T).T.
+    if is_contiguous(c, 1):
+        if is_contiguous(a, 1):
+            transb = 'n'
+            ldb = (&a[1,0]) - a0 if a.shape[0] > 1 else 1
+        elif is_contiguous(a, 0):
+            transb = 't'
+            ldb = (&a[0,1]) - a0 if a.shape[1] > 1 else 1
+        else:
+            raise ValueError("Input 'a' is neither C nor Fortran contiguous.")
```

Comment:
Grab the GIL back when you need it, instead of making the whole function require the GIL.
```
with gil:
     raise ValueError("Input 'a' is neither C nor Fortran contiguous.")
```